url5583f3dbb2_XOR19_XScript_tgz-pJ8-xscript_interactive_MainJ8/src/xscript/XUtils.java

```diff
@@ -0,0 +1,562 @@
+		}
+		set(rt, v1, s.substring(l), v2, XValue.REF_NONE);
+	}
+
+	public static String getString(XRuntime rt, XValue value) {
+		XObjectData data = getData(rt, value);
+		return data instanceof XObjectDataString ? ((XObjectDataString) data)
+				.getString() : null;
+	}
+
+	@SuppressWarnings("unchecked")
+	public static <T> T getDataAs(XRuntime rt, XValue value, Class<T> c) {
+		XObjectData data = getData(rt, value);
+		return data != null && c.isAssignableFrom(data.getClass()) ? (T) data
+				: null;
+	}
+
+	public static XObjectData getData(XRuntime rt, XValue value) {
+		XObject obj = rt.getObject(value);
+		return obj == null ? null : obj.getData();
+	}
+
+	public static boolean isInstanceOf(XRuntime rt, XValue instance, XValue type) {
+		XValue instanceType = instance.getType(rt);
+		return isDerivedOf(rt, type, instanceType);
+	}
+
+	public static boolean isDerivedOf(XRuntime rt, XValue base, XValue type) {
+		XObject typeObj = rt.getObject(type);
+		XObject baseObj = rt.getObject(base);
+		if (typeObj == null || !typeObj.isType(rt) || baseObj == null
+				|| !baseObj.isType(rt))
+			return false;
+		return ((XTypeData) typeObj.getData()).getCRO().contains(base);
+	}
+
+	public static void rethrow(Throwable e) {
+		if (e instanceof RuntimeException) {
+			throw (RuntimeException) e;
+		} else if (e instanceof Error) {
+			throw (Error) e;
+		}
+		throw new RuntimeException(e);
+	}
+
+	public static XValue[] wrap(XRuntime rt, Object[] obj) {
+		XValue[] values = new XValue[obj.length];
+		for (int i = 0; i < obj.length; i++) {
+			values[i] = wrap(rt, obj[i]);
+		}
+		return values;
+	}
+
+	public static XValue wrap(XRuntime rt, Object obj) {
+		if (obj == null) {
+			return XValueNull.NULL;
+		} else if (obj instanceof XValue) {
+			return (XValue) obj;
+		} else if (obj instanceof XObject) {
+			XObject object = (XObject) obj;
+			if (rt.getObject(object.getPointer()) != object) {
+				throw new IllegalArgumentException(
+						"XObject is form another Runtime");
+			}
+			return new XValueObj(object.getPointer());
+		} else if (obj instanceof String) {
+			return rt.alloc((String) obj);
+		} else if (obj instanceof Character) {
+			return XValueInt.valueOf(((Character) obj).charValue());
+		} else if (obj instanceof Number) {
+			if (obj instanceof Byte || obj instanceof Short
+					|| obj instanceof Integer || obj instanceof Long
+					|| obj instanceof BigInteger) {
+				return XValueInt.valueOf(((Number) obj).longValue());
+			}
+			return new XValueFloat(((Number) obj).doubleValue());
+		} else if (obj instanceof Boolean) {
+			return XValueBool.valueOf(((Boolean) obj).booleanValue());
+		} else if (obj instanceof List) {
+			List<?> l = (List<?>) obj;
+			List<XValue> list = new ArrayList<XValue>(l.size());
+			for (Object o : l) {
+				list.add(wrap(rt, o));
+			}
+			return rt.createList(list);
+		} else if (obj instanceof Map) {
+			Map<?, ?> m = (Map<?, ?>) obj;
+			Map<String, XValue> map = new HashMap<String, XValue>(m.size());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The `HashMap` constructor rehashes your hash table after 75% of all keys have been added to the table. To improve performance and prevent rehashing, use the Guava `newHashMapWithExpectedSize()` method. See the following diffs on GitHub that make this code change: [diff1](https://github.com/thefreebit/bullfrog/commit/d4d0b55af8a0eb925f039ef78d137565e1ae574b#diff-a24080844c93ebee0f466f384d2b3599cbfd502623597fc30e3364bc3fad5fedL119), [diff2](https://github.com/outbrain/ob1k/commit/5183bd681f4e9a8b67289314359b1211eb764899#diff-1891c9f8b2921c25e9406be78b9688662c7cd63b57300817fd288ed2ec0e3eeaL282)

Learn more: [HashMap constructor in the Java Platform Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/util/HashMap.html#HashMap-int-) and the [newHashMapWithExpectedSize in the Guava: Google Core Libraries for Java 23.0 API](https://guava.dev/releases/23.0/api/docs/com/google/common/collect/Maps.html#newHashMapWithExpectedSize-int-)